### PR TITLE
Refactored storage interface

### PR DESCRIPTION
**Pull request by Onager**
This is in preparation for creating new storage implementations, notably Redis.

### CODE REVIEW

File: plaso/cli/helpers/storage_format.py [hunk 1]
```diff
@@ -65,5 +75,16 @@ def ParseOptions(cls, options, configuration_object):
...
+    task_storage_format = cls._ParseStringOption(options, 'task_storage_format')
+    if not task_storage_format:
+      raise errors.BadConfigOption('Unable to determine task storage format.')
+
+    if task_storage_format not in definitions.TASK_STORAGE_FORMATS:
+      raise errors.BadConfigOption(
+          'Unsupported task storage format: {0:s}'.format(task_storage_format))
+
+    setattr(configuration_object, '_task_storage_format', task_storage_format)
+
```
Comment (joachimmetz) [inline plaso/cli/helpers/storage_format.py]: `-` white line

File: plaso/containers/tasks.py [hunk 2]
```diff
@@ -57,6 +58,7 @@ def __init__(self, session_identifier=None):
     self.session_identifier = session_identifier
     self.start_time = int(time.time() * definitions.MICROSECONDS_PER_SECOND)
     self.storage_file_size = None
+    self.storage_format = None
```
Comment (joachimmetz) [inline plaso/containers/tasks.py]: why make this part of the task?

File: plaso/containers/tasks.py [hunk 3]
```diff
@@ -35,6 +35,7 @@ class Task(interface.AttributeContainer):
     start_time (int): time that the task was started. Contains the number
         of micro seconds since January 1, 1970, 00:00:00 UTC.
     storage_file_size (int): size of the storage file in bytes.
+    storage_format (str): the format the task was stored in.
```
Comment (joachimmetz) [inline plaso/containers/tasks.py]: was stored in? or should be stored in?

is this used for informational purposes or for configuration purposes?

File: plaso/lib/errors.py [hunk 4]
```diff
@@ -79,6 +79,8 @@ class SerializationError(Error):
 class SourceScannerError(Error):
   """Class that defines source scanner errors."""
```
Comment (joachimmetz) [inline plaso/lib/errors.py]: `+` white line

File: plaso/lib/errors.py [hunk 5]
```diff
@@ -79,6 +79,8 @@ class SerializationError(Error):
...
 
+class StorageError(Error):
+  """Base class for errors relating to storage."""
```
Comment (joachimmetz) [inline plaso/lib/errors.py]: `+` white line

File: plaso/multi_processing/task_engine.py [hunk 6]
```diff
@@ -213,10 +213,10 @@ def _MergeTaskStorage(self, storage_writer):
           storage_writer.PrepareMergeTaskStorage(task)
           self._task_manager.UpdateTaskAsPendingMerge(task)
 
-      except KeyError:
+      except KeyError as exception:
         logger.error(
-            'Unable to retrieve task: {0:s} to prepare it to be merged.'.format(
-                task_identifier))
+            'Unable to retrieve task: {0:s} to prepare it to be merged. '
+            'Error: {1!s}.'.format(task_identifier, exception))
```
Comment (joachimmetz) [inline plaso/multi_processing/task_engine.py]: per style guide use `with error: {`

File: plaso/multi_processing/task_engine.py [hunk 7]
```diff
@@ -439,9 +441,10 @@ def _ScheduleTasks(self, storage_writer):
 
         if task:
           if self._ScheduleTask(task):
+            task_path_spec = task.path_spec.comparable.replace('\n', ' ')
```
Comment (joachimmetz) [inline plaso/multi_processing/task_engine.py]: task_path_spec => task_path_spec_string or something to indicate this is not a path specification object but a string representation

File: plaso/multi_processing/task_manager.py [hunk 8]
```diff
@@ -311,17 +311,21 @@ def CreateRetryTask(self):
...
-  def CreateTask(self, session_identifier):
+  def CreateTask(
+      self, session_identifier,
+      storage_format=definitions.STORAGE_FORMAT_SQLITE):
     """Creates a task.
 
     Args:
       session_identifier (str): the identifier of the session the task is
           part of.
+      storage_format (str): TODO
```
Comment (joachimmetz) [inline plaso/multi_processing/task_manager.py]: fix TODO?

File: plaso/storage/fake/writer.py [hunk 9]
```diff
@@ -210,11 +210,35 @@ def AddWarning(self, warning):
...
+    """Checks if a task is ready for merging with this session storage.
+
+    Args:
+      task (Task): task.
+
+    Returns:
+      bool: True if the task is ready to be merged.
+
+    Raises:
+      IOError: if the task storage does not exist.
```
Comment (joachimmetz) [inline plaso/storage/fake/writer.py]: does not reflect code below

File: plaso/storage/file_interface.py [hunk 10]
```diff
@@ -0,0 +1,1147 @@
...
+import shutil
+
+import abc
+import os
+import tempfile
+
+from plaso.lib import definitions
+from plaso.serializer import json_serializer
+from plaso.storage import interface
+
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: `+` white line

File: plaso/storage/file_interface.py [hunk 11]
```diff
@@ -0,0 +1,1147 @@
+# -*- coding: utf-8 -*-
+"""Storage interface classes for file-backed stores."""
+
+from __future__ import unicode_literals
+
+import shutil
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: move location of import, should be just above tempfile

File: plaso/storage/file_interface.py [hunk 12]
```diff
@@ -0,0 +1,1147 @@
...
+      IndexError: if the index is less than zero.
+    """
+    if index < 0:
+      raise IndexError(
+          'Unsupported negative index value: {0:d}.'.format(index))
+
+    if index < len(self._list):
+      return self._list[index]
+
+    return None
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: Why not raise for index > len() ?

File: plaso/storage/file_interface.py [hunk 13]
```diff
@@ -0,0 +1,1147 @@
...
+  Attributes:
+    data_size (int): total data size of the serialized attribute containers
+        on the list.
+    next_sequence_number (int): next attribute container sequence number.
+  """
+
+  def __init__(self):
+    """Initializes a serialized attribute container list."""
+    super(SerializedAttributeContainerList, self).__init__()
+    self._list = []
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: what is the purpose of `_list` ?

File: plaso/storage/file_interface.py [hunk 14]
```diff
@@ -0,0 +1,1147 @@
...
+    if index < len(self._list):
+      return self._list[index]
+
+    return None
+
+  def PopAttributeContainer(self):
+    """Pops a serialized attribute container from the list.
+
+    Returns:
+      bytes: serialized attribute container data.
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: ... or None if

File: plaso/storage/file_interface.py [hunk 15]
```diff
@@ -0,0 +1,1147 @@
...
+    if self._serializers_profiler:
+      self._serializers_profiler.StartTiming(
+          attribute_container.CONTAINER_TYPE)
+
+    try:
+      attribute_container_data = self._serializer.WriteSerialized(
+          attribute_container)
+      if not attribute_container_data:
+        raise IOError(
+            'Unable to serialize attribute container: {0:s}.'.format(
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: `+` with error

File: plaso/storage/file_interface.py [hunk 16]
```diff
@@ -0,0 +1,1147 @@
...
+    Raises:
+      IOError: when the storage file is closed or read-only.
+      OSError: when the storage file is closed or read-only.
+    """
+    if not self._is_open:
+      raise IOError('Unable to write to closed storage file.')
+
+    if self._read_only:
+      raise IOError('Unable to write to read-only storage file.')
+
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: `+` white line

File: plaso/storage/file_interface.py [hunk 17]
```diff
@@ -0,0 +1,1147 @@
...
+    Returns:
+      str: the serialization format, or None if not available.
+    """
+    if self._storage_file:
+      return self._storage_file.serialization_format
+
+    return None
+
+  def GetStorageType(self):
+    """Retrieves the storage type of the underlying storage file.
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: Why is this relevant to expose? The idea of the reader is to hide the fact the storage is a file or something else

File: plaso/storage/file_interface.py [hunk 18]
```diff
@@ -0,0 +1,1147 @@
...
+    Returns:
+      int: the format version, or None if not available.
+    """
+    if self._storage_file:
+      return self._storage_file.format_version
+
+    return None
+
+  def GetSerializationFormat(self):
+    """Retrieves the serialization format of the underlying storage file.
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: Why is this relevant to expose? The idea of the reader is to hide the fact the storage is a file or something else

File: plaso/storage/interface.py [hunk 19]
```diff
@@ -4,172 +4,237 @@
...
 
-class SerializedAttributeContainerList(object):
-  """Serialized attribute container list.
-
-  The list is unsorted and pops attribute containers in the same order as
-  pushed to preserve order.
-
-  The GetAttributeContainerByIndex method should be used to read attribute
-  containers from the list while it being filled.
+# pylint: disable=redundant-returns-doc,redundant-yields-doc
```
Comment (joachimmetz) [inline plaso/storage/interface.py]: move these overrides into the class

File: plaso/storage/interface.py [hunk 20]
```diff
@@ -4,172 +4,237 @@
...
+class BaseStore(object):
+  """Storage interface.
 
   Attributes:
-    data_size (int): total data size of the serialized attribute containers
-        on the list.
-    next_sequence_number (int): next attribute container sequence number.
+    format_version (int): storage format version.
+    serialization_format (str): serialization format.
+    storage_type (str): storage type.
```
Comment (joachimmetz) [inline plaso/storage/interface.py]: what is the rational to make the formats part of the interface?

**Comment by Onager** [comment]
Opened https://github.com/log2timeline/plaso/pull/2650 to move the files first.

File: plaso/storage/file_interface.py (hunk 18 above)
Comment (Onager) [reply plaso/storage/file_interface.py]: This is only used by pinfo, to display information about the storage details of the store.

File: plaso/storage/file_interface.py (hunk 17 above)
Comment (Onager) [reply plaso/storage/file_interface.py]: This is only used by pinfo, to display information about the storage details of the store.

File: plaso/storage/interface.py (hunk 19 above)
Comment (Onager) [reply plaso/storage/interface.py]: Done

File: plaso/storage/interface.py (hunk 20 above)
Comment (Onager) [reply plaso/storage/interface.py]: These haven't moved, hopefully the cleaner diff makes this easier to see.

File: plaso/storage/file_interface.py (hunk 14 above)
Comment (Onager) [reply plaso/storage/file_interface.py]: Done

File: plaso/storage/file_interface.py (hunk 13 above)
Comment (Onager) [reply plaso/storage/file_interface.py]: This class was just moved, but there doesn't seem to be another list that this would get confused with other than the class itself.

File: plaso/storage/file_interface.py (hunk 12 above)
Comment (Onager) [reply plaso/storage/file_interface.py]: I don't know - this class was just moved.

File: plaso/containers/tasks.py (hunk 2 above)
Comment (joachimmetz) [reply plaso/containers/tasks.py]: ack, per conversation about this

File: plaso/multi_processing/task_manager.py [hunk 21]
```diff
@@ -311,17 +311,22 @@ def CreateRetryTask(self):
...
-  def CreateTask(self, session_identifier):
+  def CreateTask(
+      self, session_identifier,
+      storage_format=definitions.STORAGE_FORMAT_SQLITE):
     """Creates a task.
 
     Args:
       session_identifier (str): the identifier of the session the task is
           part of.
+      storage_format (str): the storage format that the task should be stored
```
Comment (joachimmetz) [inline plaso/multi_processing/task_manager.py]: => `(Optional[str])`

File: plaso/storage/file_interface.py (hunk 18 above)
Comment (joachimmetz) [reply plaso/storage/file_interface.py]: ack, per conversation

File: plaso/storage/file_interface.py (hunk 17 above)
Comment (joachimmetz) [reply plaso/storage/file_interface.py]: ack, per conversation

File: plaso/storage/file_interface.py [hunk 22]
```diff
@@ -895,15 +798,19 @@ def FinalizeTaskStorage(self, task):
     if self._storage_type != definitions.STORAGE_TYPE_SESSION:
       raise IOError('Unsupported storage type.')
 
-    storage_file_path = self._GetTaskStorageFilePath(task)
-    processed_storage_file_path = self._GetProcessedStorageFilePath(task)
+    if not task.storage_format in definitions.STORAGE_FORMATS:
+      raise IOError('Unsupported storage format')
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: add "Unsupported storage format" as a reason to raise to the docstring

File: plaso/storage/file_interface.py [hunk 23]
```diff
@@ -994,6 +901,9 @@ def RemoveProcessedTaskStorage(self, task):
     if self._storage_type != definitions.STORAGE_TYPE_SESSION:
       raise IOError('Unsupported storage type.')
 
+    if task.storage_format != definitions.STORAGE_FORMAT_SQLITE:
+      raise IOError('Unsupported storage format.')
```
Comment (joachimmetz) [inline plaso/storage/file_interface.py]: add "Unsupported storage format" as a reason to raise to the docstring

File: plaso/storage/interface.py (hunk 20 above)
Comment (joachimmetz) [reply plaso/storage/interface.py]: ack, staged changes help in clarification

File: plaso/storage/interface.py [hunk 24]
```diff
@@ -25,68 +63,177 @@ def __init__(self):
...
+    """Closes the store."""
 
-  @abc.abstractmethod
   def GetAnalysisReports(self):
     """Retrieves the analysis reports.
 
-    Yields:
-      AnalysisReport: analysis report.
+    Returns:
+      generator(AnalysisReport): analysis report generator.
```
Comment (joachimmetz) [inline plaso/storage/interface.py]: use `[]` instead of `()`

File: plaso/storage/interface.py [hunk 25]
```diff
@@ -122,23 +271,40 @@ def GetEventTagByIdentifier(self, identifier):
...
 
-  @abc.abstractmethod
+  # pylint: disable=redundant-yields-doc
   def GetEventTags(self):
     """Retrieves the event tags.
 
     Yields:
       EventTag: event tag.
     """
+    self._GetAttributeContainers(self._CONTAINER_TYPE_EVENT_TAG)
```
Comment (joachimmetz) [inline plaso/storage/interface.py]: missing returns

File: plaso/storage/interface.py [hunk 26]
```diff
@@ -122,23 +271,40 @@ def GetEventTagByIdentifier(self, identifier):
...
+
+    Raises:
+      OSError: if an invalid identifier is provided.
+      IOError: if an invalid identifier is provided.
     """
+    return self._GetAttributeContainerByIdentifier(
+        self._CONTAINER_TYPE_EVENT_TAG, identifier)
 
-  @abc.abstractmethod
+  # pylint: disable=redundant-yields-doc
```
Comment (joachimmetz) [inline plaso/storage/interface.py]: Remove override, this method should return a generator

File: plaso/multi_processing/task_manager.py (hunk 21 above)
Comment (Onager) [reply plaso/multi_processing/task_manager.py]: Done

File: plaso/storage/file_interface.py (hunk 22 above)
Comment (Onager) [reply plaso/storage/file_interface.py]: Done

File: plaso/storage/file_interface.py (hunk 23 above)
Comment (Onager) [reply plaso/storage/file_interface.py]: Done

File: plaso/storage/interface.py (hunk 26 above)
Comment (Onager) [reply plaso/storage/interface.py]: Done

File: plaso/storage/interface.py (hunk 25 above)
Comment (Onager) [reply plaso/storage/interface.py]: Done

File: plaso/multi_processing/task_engine.py [hunk 27]
```diff
@@ -423,7 +423,9 @@ def _ScheduleTasks(self, storage_writer):
           task = self._task_manager.CreateRetryTask()
 
         if not task and event_source:
-          task = self._task_manager.CreateTask(self._session_identifier)
+          task = self._task_manager.CreateTask(
+              self._session_identifier,
+              self._processing_configuration.task_storage_format)
```
Comment (joachimmetz) [inline plaso/multi_processing/task_engine.py]: => `storage_format=self._processing_configuration.task_storage_format`

File: plaso/storage/interface.py [hunk 28]
```diff
@@ -96,24 +243,26 @@ def GetEventDataByIdentifier(self, identifier):
...
+        self._CONTAINER_TYPE_EVENT_DATA, identifier)
 
-  @abc.abstractmethod
   def GetEvents(self):
     """Retrieves the events.
 
     Yields:
       EventObject: event.
     """
+    for event in self._GetAttributeContainers(self._CONTAINER_TYPE_EVENT):
```
Comment (joachimmetz) [inline plaso/storage/interface.py]: Why not `return self._GetAttributeContainers(self._CONTAINER_TYPE_EVENT)` ?

**Comment by joachimmetz** [comment]
Please resolve merge conflicts
